SitumWayfinding/Classes/Positioning/Internal/MarkerRenderer.swift

```diff
@@ -22,87 +22,114 @@ class MarkerRenderer {
             markers.append(marker)
-            marker.setMapView(mapView: mapView)
+            insertMarkerInGoogleMaps(marker: marker)
         }
     }
     
-    func displayDestinationMarker(marker: SitumMarker, forFloor floor: SITFloor) {
-        cleanMarkers()
+    func displayOnlyDestinationMarker(_ marker: SitumMarker, forFloor floor: SITFloor) {
+        removeMarkers()
         if floor.identifier == marker.floorIdentifier {
             markers.append(marker)
-            loadSelectedIcon(forMarker: marker)
+            selectMarker(marker)
         }
     }
     
-    private func cleanMarkers() {
+    private func removeMarkers() {
         for marker in markers {
             // disassociate markers from map (otherwise we lost reference and they get stuck in map forever)
-            marker.setMapView(mapView: nil)
+            removeMarkerFromGoogleMaps(marker: marker)
         }
         markers.removeAll()
     }
     
-    func loadSelectedIcon(forMarker marker: SitumMarker) {
+    func selectMarker(_ marker: SitumMarker) {
         if marker.isPoiMarker {
-            loadIcon(forMarker: marker, selected: true)
+            loadIcon(forMarker: marker, selected: true) { [weak self] marker in
+                self?.selectMarkerInGoogleMaps(marker: marker)
+            }
         }
         if marker.isCustomMarker {
-            mapView.selectedMarker = marker.gmsMarker
-            if marker.gmsMarker.map == nil {
-                marker.setMapView(mapView: mapView)
-            }
+            selectMarkerInGoogleMaps(marker: marker)
         }
     }
     
-    func loadUnselectedIcon(forMarker marker: SitumMarker) {
+    func deselectMarker(_ marker: SitumMarker) {
         if marker.isPoiMarker {
-            loadIcon(forMarker: marker, selected: false)
+            loadIcon(forMarker: marker, selected: false) { [weak self] marker in
+                self?.deselectMarkerFromGoogleMaps(marker: marker)
+            }
         }
-        // when unselected a custom marker we remove it
         if marker.isCustomMarker {
-            marker.setMapView(mapView: nil)
+            removeMarkerFromGoogleMaps(marker: marker)
             markers = markers.filter { element in element != marker }
         }
+    }
+    
+    private func selectMarkerInGoogleMaps(marker: SitumMarker) {
+        insertMarkerInGoogleMaps(marker: marker)
+        mapView.selectedMarker = marker.gmsMarker
+    }
+    
+    private func insertMarkerInGoogleMaps(marker: SitumMarker) {
+        if marker.gmsMarker.map == nil {
+            marker.setMapView(mapView: mapView)
+        }
+    }
+    
+    private func removeMarkerFromGoogleMaps(marker: SitumMarker) {
+        if mapView.selectedMarker == marker.gmsMarker {
+            deselectMarkerFromGoogleMaps(marker: marker)
+        }
+        marker.setMapView(mapView: nil)
+    }
+    
+    private func deselectMarkerFromGoogleMaps(marker: SitumMarker) {
         mapView.selectedMarker = nil
```

Comment:
Creo que con estos métodos quedan máis claros e limpos os métodos públicos que os usan. Ademáis creo que nos empeza a dar unha pista de que métodos vamos a necesitar a nivel interfaz máis adiante. 
Non sei como o ves, sobre todo para a función que solo ten a liña `mapView.selectedMarker = nil`, igual demasiado verboso tantos métodos pero creo que quedan moito máis semánticos.

Reply:
A min gustame, porque indicanos moi ben onde imos estando acoplados con googleMaps, o que nos permitira refactorizar de forma máis doada máis adiante se queremos engadir outro proveedor de mapas